### DISCUSSION TITLE
line wrap issue in git-bash

### DISCUSSION THREAD

**Discussion by ProgRick** [JanDeDobbeleer/oh-my-posh#1436] [Troubleshoot]
I'm using oh-my-posh in git-hub and there is an issue where my command line text is too long for a single line, instead of wrapping onto a new line, it wraps back over itself. Is there a way to fix this?

The version of oh-my-posh I have installed is 6.27.1
![line-wrap](https://user-images.githubusercontent.com/96149739/146085446-9e302b56-4c44-4531-8fb9-a9c652d73571.jpg)

**Comment by JanDeDobbeleer** [comment]
What do you mean with "in github"?

**Reply by ProgRick** [reply]
Sorry I meant git-bash.

**Reply by JanDeDobbeleer** [reply]
@ProgRick you need to update git, seems like there's an issue on older versions.

**Reply by Helvette** [reply]
I have this problem in Windows Terminal (using Git Bash) but in Git Bash it wraps correctly. I can't find a solution for the Windows Terminal :/

**Reply by JanDeDobbeleer** [reply]
@Helvette I would ask the question in the Windows Terminal repo. Maybe they can help. The terminal shouldn't influence this behavior.

**Reply by ProgRick** [reply]
> @ProgRick you need to update git, seems like there's an issue on older versions.

@JanDeDobbeleer I know its been a while, but I think I have some more info. The issue is in git-bash running on Windows. I just got a new laptop and reinstalled everything and the problem still happens. I have oh-my-posh configured in both Windows Powershell and git-bash, running the exactly same installation of oh-my-posh. When a line needs to wrap in Powershell it does it properly, dropping down a line. When a line wraps in git-bash the cursor starts over at the beginning of the line that I am already on.

I found this person with the same behavior trying to do something very similar to what oh-my-posh is doing: https://stackoverflow.com/questions/18559417/bash-prompt-line-wrapping-issue
I think the second answer might be key. It starts with "The wrapping error occurs whenever a non-printing character is not escaped". I believe the PS1 variable is actually set inside of the oh-my-posh.exe file, so I can't immediately determine if that is the case. I could check the source for the project, but I might not be able to understand what I am looking at without lots of time.

**Reply by JanDeDobbeleer** [reply]
@ProgRick we escape all non-printing characters as otherwise the prompt would act really strange. But i'll have s look with your config if you would be so kind to provide it.

**Comment by ghost** [comment]
@ProgRick @JanDeDobbeleer Is there any update on this issue?

**Reply by JanDeDobbeleer** [reply]
Not at all.

**Reply by ghost** [reply]
I'm facing the same issue, and I'm also using oh-my-posh with git-bash in the Windows terminal app. Let me know if you need anything to look into this issue.

**Reply by ProgRick** [reply]
FWIW, I am no longer using a Windows machine for work. I'm now using zsh on a Mac now and I don't have this issue. I think it was something specific to git-bash on Windows.

**Reply by JanDeDobbeleer** [reply]
@dhairya-niq I would need you config to see if I can reproduce this.

**Reply by ghost** [reply]
Sure @JanDeDobbeleer , I've attached my JSON configuration, and I'm using the below prompt.

`eval "$(oh-my-posh init bash --config C:/Users/<USER_NAME>/Documents/omp_themes/tokyonight_storm.omp.json)"`

[tokyonight_storm.omp.json](https://github.com/JanDeDobbeleer/oh-my-posh/files/14235713/tokyonight_storm.omp.json)

**Reply by JanDeDobbeleer** [reply]
Still have no idea what's causing this, but it's not reproducible on my end.

![image](https://github.com/JanDeDobbeleer/oh-my-posh/assets/2492783/b4abddc9-31ed-4871-9532-98f416bf7404)

**Reply by ghost** [reply]
Here is an update. It seems like this issue is only limited to Windows 10.
I tried with my other machine running on Windows 11, and it's working as expected.

**Reply by JanDeDobbeleer** [reply]
@dhairya-niq in that case probably a conhost issue, not something I can solve.

**Reply by ghost** [reply]
Got it. Thanks

**Comment by rquadling** [comment]
Sort of related ... ish ...

I'm on MacOS and having the exact same issue with git-prompt. (which I'm hoping to be able to find and fix).

A way to identify the issue is to run the `env` command.

Now ... admittedly ... my prompt has a LOT of data on it and I've been hacking away trying to work out why I was having the early wrapping .. and then suddenly not when testing the amendments. I was sometimes in a git checkout. And sometimes not.

If I run my `env` command, for the `PS1` envvar I get ...
```
PS1=\001$(code=${?##0};echo ${code:+$(tput setaf 9)\002(${code})\ })\001$(tput sgr0)\002\001$(tput setaf 6)\002${PS1_SHELL_IN}\001$(tput setaf 14)\002$SHLVL\001$(tput setaf 6)\002${PS1_SHELL_OUT} \001$(tput sgr0)\002\001$(tput setaf 5)\002\D{%Y-%m-%d %H:%M:%S} \001$(tput sgr0)\002\001$(tput setaf 6)\002\u\001$(tput sgr0)\002\001$(tput setaf 8)\002@\001$(tput sgr0)\002\001$(tput setaf 2)\002\h\001$(tput sgr0)\002\001$(tput setaf 8)\002:\001$(tput sgr0)\002\001$(tput setaf 3)\002\w\001$(tput sgr0)\002\001$([[ $(dirs -p | wc -l) -gt 1 ]] && echo "\002 [\001$(dirs -p | wc -l | xargs)\001]\002") (${__git_ps1_branch_name} *=) \\$ 
```
But what's missing in that is the `${__git_ps1_branch_name}` is in green and the `*` is red, matching the display as shown in the actual prompt ...
```
... ~/dt/stack (master *=) $ env
```
Here, `master` is in green and the `*` is red.

<img width="1259" alt="image" src="https://github.com/JanDeDobbeleer/oh-my-posh/assets/12801/735f6ae4-fcec-470a-9cb9-c7e61f56b26a">
and <img width="244" alt="image" src="https://github.com/JanDeDobbeleer/oh-my-posh/assets/12801/0ee2d86b-baf4-48b1-b162-f880d360bbb6">

This is the way I've found what was causing the wrapping.

It may be of use to you all.

**Comment by rquadling** [comment]
And my fix for this was to edit the file git-prompt.sh

And add an escape `\` in the following lines:

```
		local c_red=$'\\001\e[31m\\002'
		local c_green=$'\\001\e[32m\\002'
		local c_lblue=$'\\001\e[1;34m\\002'
		local c_clear=$'\\001\e[0m\\002'
```
for the `\001` and `\002` becoming `\\001` and `\\002`.

https://github.com/git/git/commit/7ee1af8cb8b97385fc4e603bc024d877def5adb4#r139625087

**Reply by JosueOrtiz-fpga** [reply]
Very helpful!

In my case, I open a Git Bash terminal within VSCODE and was having a similar issue of the current line being overwritten (visually) instead of wrapping to a new line. I was missing a  backslash before the '$' at the end of my PS1 variable.

Before:
`PS1='\e[0;32m\w\e[1;33m$(__git_ps1):\e[m\$'`
After:
`PS1='\e[0;32m\w\e[1;33m$(__git_ps1):\e[m\\$'`
